### PR TITLE
obs-qsv11: Add HEVC support

### PR DESCRIPTION
**Pull request by yuriy-chumak**
### Description
- Add Intel HEVC hardware encoder (QSV H.265).
~~- Add *Simple Output Mode* for Intel HEVC hardware encoder.~~

### Motivation and Context
- We should have a way to use Intel HEVC hardware encoder in streaming and recording.
~~- We should have easy way (*Simple Output Mode*) to use Intel HEVC hardware encoder in streaming and recording.~~

### How Has This Been Tested?
MSVC debugger, VLC color/metadata checks
- [x] H.264, NV12, 709 -> Regular
- [x] H.264, NV12, 2100PQ -> HDR on 8-bit fail
- [x] H.264, NV12, 2100HLG -> HDR on 8-bit fail
- [x] H.264, I420, 709 -> NV12 fallback
- [x] H.264, I420, 2100PQ -> HDR on 8-bit fail
- [x] H.264, I420, 2100HLG -> HDR on 8-bit fail
- [x] H.264, I010, 709 -> 10-bit fail
- [x] H.264, I010, 2100PQ -> 10-bit fail
- [x] H.264, I010, 2100HLG -> 10-bit fail
- [x] H.264, P010, 709 -> 10-bit fail
- [x] H.264, P010, 2100PQ -> 10-bit fail
- [x] H.264, P010, 2100HLG -> 10-bit fail
- [x] HEVC, NV12, 709 -> Regular
- [x] HEVC, NV12, 2100PQ -> HDR on 8-bit fail
- [x] HEVC, NV12, 2100HLG -> HDR on 8-bit fail
- [x] HEVC, I420, 709 -> NV12 fallback
- [x] HEVC, I420, 2100PQ -> HDR on 8-bit fail
- [x] HEVC, I420, 2100HLG -> HDR on 8-bit fail
- [x] HEVC, I010, 709 -> P010 fallback
- [x] HEVC, I010, 2100PQ -> P010 fallback
- [x] HEVC, I010, 2100HLG -> P010 fallback
- [x] HEVC, P010, 709 -> Regular
- [x] HEVC, P010, 601 -> Regular
- [x] HEVC, P010, sRGB -> Regular
- [x] HEVC, P010, 2100PQ -> Regular
- [x] HEVC, P010, 2100HLG -> Regular

RenderDoc color check
- [x] H.264, 709, limited vs. full range

MediaInfo metadata checks:
- [x] H.264, 709, Limited range
- [x] H.264, 709, Full range
- [x] HEVC, 709, no display mastering metadata, no content light metadata (garbage content light level, but not our fault)
- [x] HEVC, 601, no display mastering metadata, no content light metadata (garbage content light level, but not our fault)
- [x] HEVC, sRGB, no display mastering metadata, no content light metadata (garbage content light level, but not our fault)
- [x] HEVC, 2100PQ, display mastering metadata, content light metadata (garbage content light level, but not our fault)
- [x] HEVC, 2100HLG, display mastering metadata, content light metadata (garbage content light level, but not our fault)

Compile/UI check:
- [x] Verify HEVC encoder not available when ENABLE_HEVC is disabled

Chrome checks:
- [x] Test HEVC 709 against YouTube streaming
- [x] Test HEVC 2100PQ against YouTube streaming
- [x] Test HEVC 2100HLG against YouTube streaming

### Types of changes
- New feature (non-breaking change which adds functionality)

### Checklist:
- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

### CODE REVIEW

**Comment by gxalpha** [comment]
@yuriy-chumak marking as draft as this PR depends on another PR (#6521). This makes sure we don't accidentally merge it before the other one.
Feel free to mark as ready for review once #6521 has been merged and this PR has subsequently been rebased.

**Comment by yuriy-chumak** [comment]
Looking back at the [comment](https://github.com/obsproject/obs-studio/pull/6521#issuecomment-1135439238), I removed *Simple Output Mode* from the PR. Simple Output Mode will be provided in additional PR, if needed.

Additionally, I removed PR dependency from the PR #6521.

So, pull request is ready to review. I'm removing Draft mark.

File: plugins/obs-qsv11/obs-qsv265.c
Comment (jpark37) [inline plugins/obs-qsv11/obs-qsv265.c:158]: Remove "Encoder" for consistency with H.264.

File: plugins/obs-qsv11/CMakeLists.txt
Comment (jpark37) [inline plugins/obs-qsv11/CMakeLists.txt:85]: Can you fold obs-qsv265.c into obs-qsv11.c to reduce copy/paste?

File: plugins/obs-qsv11/obs-qsv11-plugin-main.c
Comment (jpark37) [inline plugins/obs-qsv11/obs-qsv11-plugin-main.c:64]: "Intel Quick Sync Video encoder support for Windows" and remove #ifdef.

File: plugins/obs-qsv11/common_directx11.cpp
Comment (jpark37) [inline plugins/obs-qsv11/common_directx11.cpp:229]: Have we had this wrong the whole time? Was this caught by the D3D debug layer?

File: libobs/obs-hevc.h
Comment (jpark37) [inline libobs/obs-hevc.h:33]: Do not expose enum in header if it's not public.

File: libobs/obs-hevc.h
Comment (jpark37) [inline libobs/obs-hevc.h:41]: Do not expose enum in header if it's not public.

File: plugins/obs-qsv11/CMakeLists.txt
Comment (jpark37) [inline plugins/obs-qsv11/CMakeLists.txt:86]: Were these files copy/pasted from the Intel SDK? Otherwise follow obs file naming, e.g. lower_case_words.cpp.

**Review by jpark37 — CHANGES_REQUESTED** [review]
Initial review feedback items. I'll look in more detail after they are resolved.

Thanks.

File: plugins/obs-qsv11/obs-qsv265.c
Comment (jpark37) [inline plugins/obs-qsv11/obs-qsv265.c:504]: Please support SDR. Currently seeing green screen in video using NV12 & Rec. 709.

File: plugins/obs-qsv11/CMakeLists.txt
Comment (jpark37) [reply plugins/obs-qsv11/CMakeLists.txt:86]: Never mind. I see the other files. Looks like they were provided to OBS from Intel

File: plugins/obs-qsv11/QSV_HevcEncoder.h
Comment (jpark37) [inline plugins/obs-qsv11/QSV_HevcEncoder.h:76]: Use "main" and "main10" to match FFmpeg style:
https://github.com/FFmpeg/FFmpeg/blob/415f012359364a77e8394436f222b74a8641a3ee/libavcodec/qsvenc_hevc.c#L251

File: plugins/obs-qsv11/obs-qsv265.c
Comment (jpark37) [inline plugins/obs-qsv11/obs-qsv265.c:196]: Default to "main" but dynamically choose MFX_PROFILE_HEVC_MAIN10 if the bit-depth is 10. (Same logic that NVENC uses.)

File: plugins/obs-qsv11/obs-qsv265.c
Comment (jpark37) [inline plugins/obs-qsv11/obs-qsv265.c:364]: Not sure what this does, or why it's necessary.

File: plugins/obs-qsv11/common_directx11.cpp
Comment (jpark37) [reply plugins/obs-qsv11/common_directx11.cpp:229]: Fwiw, Intel still uses D3D11_BIND_DECODER in their sample. Do they have a bug?
https://github.com/Intel-Media-SDK/MediaSDK/blob/34ec1df8448c1f26dfd25a5fa54f4c0486d479b7/tutorials/common/common_directx11.cpp#L225

File: plugins/obs-qsv11/obs-qsv265.c
Comment (jpark37) [inline plugins/obs-qsv11/obs-qsv265.c:195]: Default for H.264 is 3000. Why change to 30000 for HEVC?

File: plugins/obs-qsv11/obs-qsv265.c
Comment (jpark37) [inline plugins/obs-qsv11/obs-qsv265.c:197]: Default for H.264 is "CBR". Why change to "VBR" for HEVC?

File: plugins/obs-qsv11/obs-qsv265.c
Comment (jpark37) [inline plugins/obs-qsv11/obs-qsv265.c:206]: Default for H.264 is 3. Why change to 2 for HEVC?

File: plugins/obs-qsv11/obs-qsv265.c
Comment (jpark37) [inline plugins/obs-qsv11/obs-qsv265.c:207]: Default for H.264 is "normal". Why change to "ultra-low" for HEVC?

File: plugins/obs-qsv11/obs-qsv265.c
Comment (jpark37) [inline plugins/obs-qsv11/obs-qsv265.c:208]: Default for H.264 is 3. Why change to 0 for HEVC?

File: plugins/obs-qsv11/CMakeLists.txt
Comment (jpark37) [reply plugins/obs-qsv11/CMakeLists.txt:85]: I also think the QSV_HevcEncoder files should be folded into the original QSV_Encoder files.

File: plugins/obs-qsv11/obs-qsv265.c
Comment (yuriy-chumak) [reply plugins/obs-qsv11/obs-qsv265.c:158]: Done.

File: plugins/obs-qsv11/CMakeLists.txt
Comment (yuriy-chumak) [reply plugins/obs-qsv11/CMakeLists.txt:85]: in progress..

File: plugins/obs-qsv11/obs-qsv11-plugin-main.c
Comment (yuriy-chumak) [reply plugins/obs-qsv11/obs-qsv11-plugin-main.c:64]: Done.

File: plugins/obs-qsv11/common_directx11.cpp
Comment (yuriy-chumak) [reply plugins/obs-qsv11/common_directx11.cpp:229]: Yes, I think so. A Bug.
https://docs.microsoft.com/en-us/windows/win32/api/d3d11/ne-d3d11-d3d11_bind_flag

D3D11_BIND_DECODER: The common way to create resources for a **decoder output** is by calling the ID3D11Device::CreateTexture2D method to create an array of 2D textures.

D3D11_BIND_VIDEO_ENCODER: The common way to create **resources for a video encoder** is by calling the ID3D11Device::CreateTexture2D method to create an array of 2D textures.

File: libobs/obs-hevc.h
Comment (yuriy-chumak) [reply libobs/obs-hevc.h:41]: Done.

File: libobs/obs-hevc.h
Comment (yuriy-chumak) [reply libobs/obs-hevc.h:33]: Done.

File: plugins/obs-qsv11/QSV_HevcEncoder.h
Comment (yuriy-chumak) [reply plugins/obs-qsv11/QSV_HevcEncoder.h:76]: Done.

File: plugins/obs-qsv11/obs-qsv265.c
Comment (yuriy-chumak) [reply plugins/obs-qsv11/obs-qsv265.c:364]: This is artifact of implementation, where we disable HDR support if no hardware encoder found.

File: plugins/obs-qsv11/obs-qsv265.c
Comment (yuriy-chumak) [reply plugins/obs-qsv11/obs-qsv265.c:195]: 4K / 60 fps. In fact we should adapt this value depending on resolution and fps, but I think this should be another task with deeply investigation of appropriate values.

For example, [from](https://www.openreel.com/blog/bitrate-video-solutions-overview/) (last one row):
image

Comment (yuriy-chumak) [reply plugins/obs-qsv11/obs-qsv265.c:195]: Same [for YouTube](https://support.google.com/youtube/answer/2853702?hl=en#zippy=%2Ck-p-fps):
image

File: plugins/obs-qsv11/obs-qsv265.c
Comment (yuriy-chumak) [reply plugins/obs-qsv11/obs-qsv265.c:197]: I tried CBR and VBR for HEVC/HDR (using YouTube as streaming platform) and got better results for VBR (in sense of CPU usage/network utilization) for 4K/60.
For sure, this is not absolutely correct answer because it depends of other options and is not . But my current point of view is VBR.

So, leave it as is or change to?

File: plugins/obs-qsv11/obs-qsv265.c
Comment (yuriy-chumak) [reply plugins/obs-qsv11/obs-qsv265.c:206]: Because it is not working for 3.
Exactly: "it's not working for more than 2 on the specific tested PC with specific tested drivers". And, and think, I saw the limitations...

I'll recheck and change.

File: plugins/obs-qsv11/obs-qsv265.c
Comment (yuriy-chumak) [reply plugins/obs-qsv11/obs-qsv265.c:207]: For my point-of-view ultra-low for HEVC works better than normal, but not a lot.
I'll change to normal.

File: plugins/obs-qsv11/obs-qsv265.c
Comment (yuriy-chumak) [reply plugins/obs-qsv11/obs-qsv265.c:208]: Same situation like NVENC. Year ago NVidia drivers fails HEVC encoding for non zero "bframes". Latest drivers accept higher values (you helped me with drivers, thanks).

My current Intel is not working with "more than 0". And it's no good to see "fail, please update drivers" with default HEVC configuration.

I propose to leave "bframes" as "0".

File: plugins/obs-qsv11/CMakeLists.txt
Comment (yuriy-chumak) [reply plugins/obs-qsv11/CMakeLists.txt]: I merged changed as much as possible. Looks like further merging will lead to bad readability of the code.

File: plugins/obs-qsv11/obs-qsv265.c
Comment (yuriy-chumak) [reply plugins/obs-qsv11/obs-qsv265.c]: This code only differs PQ from the HLG.
I'll check...

File: plugins/obs-qsv11/common_directx11.cpp
Comment (jpark37) [reply plugins/obs-qsv11/common_directx11.cpp:229]: I know from the name, it looks like a bug, but maybe Intel has a special reason. I'll get separate clarification from them.

File: plugins/obs-qsv11/CMakeLists.txt
Comment (jpark37) [reply plugins/obs-qsv11/CMakeLists.txt]: Ok, I'll take over from here.

File: plugins/obs-qsv11/obs-qsv265.c
Comment (jpark37) [reply plugins/obs-qsv11/obs-qsv265.c]: It sounds like for consistency, we should use a value <= H.264 since HEVC is more efficient. if we want to overhaul default bitrates, that's a separate task that should span all encoders.

File: plugins/obs-qsv11/obs-qsv265.c
Comment (jpark37) [reply plugins/obs-qsv11/obs-qsv265.c]: This is Twitch's reason for preferring CBR to VBR in a streaming scenario, and it probably applies to YouTube streaming too:
> You may have noticed the option to select between VBR (variable bitrate) and CBR (constant bitrate). Twitch suggests that all Broadcasters use CBR for several reasons, all of which relate to the final quality of service (QoS) that your viewers will experience.

> The main problem with VBR is that during lulls in the action: paused games, hero selection screens, etc, VBR streams produce a significantly lower bitrate. This can cause issues (stream buffering, dropped frames) with many end-user connections when the bitrate spikes back up during the action (team fights, high motion scenes).

> VBR also leads to issues when sending data to the Twitch network over many ISPs. When connecting to Twitch, the route your ISP takes may not be stable enough to handle sudden increases in bandwidth, leading to “broadcast starvation.”

> Broadcast starvation (instability) is when video data does not arrive in a timely fashion or is completely missing. This happens when there are network problems between you and Twitch, when you attempt to stream at a bitrate that is too high for your network/ISP, or have problems with your router causing frames to be dropped before entering the Twitch network. This leads to lag for everyone (including people watching on your lower resolutions), buffering, and stream disconnections.

File: plugins/obs-qsv11/obs-qsv265.c
Comment (jpark37) [reply plugins/obs-qsv11/obs-qsv265.c]: I'll test too.

File: plugins/obs-qsv11/obs-qsv265.c
Comment (jpark37) [reply plugins/obs-qsv11/obs-qsv265.c]: I'll test too.

**Comment by jpark37** [comment]
I responded to everything I could respond to. If you want to leave the branch to me as it is, I'm fine with that.

**Comment by yuriy-chumak** [comment]
I find issue with profile changing after merge. So fixed and pushed.
Code tested and ready to review.

> I responded to everything I could respond to. If you want to leave the branch to me as it is, I'm fine with that.
I see.
Yes, thank you. I'm leaving PR to you.

In case of any additional questions to me please ask with "@" - I'll answer for sure.

File: plugins/obs-qsv11/obs-qsv265.c
Comment (jpark37) [reply plugins/obs-qsv11/obs-qsv265.c]: Rec. 709 seems to be working now.

File: plugins/obs-qsv11/obs-qsv265.c
Comment (jpark37) [reply plugins/obs-qsv11/obs-qsv265.c]: Automatically switches to main10 if P010 is active now.

File: plugins/obs-qsv11/obs-qsv265.c
Comment (jpark37) [reply plugins/obs-qsv11/obs-qsv265.c]: Looks like this is gone now.

File: plugins/obs-qsv11/obs-qsv265.c
Comment (jpark37) [reply plugins/obs-qsv11/obs-qsv265.c]: I switched the settings back to match H.264 to start with. We can optimize from there.

File: plugins/obs-qsv11/obs-qsv265.c
Comment (jpark37) [reply plugins/obs-qsv11/obs-qsv265.c]: Reverting to default for now.

File: plugins/obs-qsv11/obs-qsv265.c
Comment (jpark37) [reply plugins/obs-qsv11/obs-qsv265.c]: Reverting to default for now.

File: plugins/obs-qsv11/obs-qsv265.c
Comment (jpark37) [reply plugins/obs-qsv11/obs-qsv265.c]: Reverting to default for now.

File: plugins/obs-qsv11/obs-qsv265.c
Comment (jpark37) [reply plugins/obs-qsv11/obs-qsv265.c]: Not seeing an issue recording with keyint_sec=3. Reverting for now.

File: plugins/obs-qsv11/CMakeLists.txt
Comment (jpark37) [reply plugins/obs-qsv11/CMakeLists.txt]: I combined the files. 700 line diff instead of 2000.

File: plugins/obs-qsv11/common_directx11.cpp
Comment (jpark37) [reply plugins/obs-qsv11/common_directx11.cpp]: Seems like D3D11_BIND_VIDEO_ENCODER is being omitted on purpose. This is what Intel's code looks like, which I've copied into the PR:
```
#if (MFX_VERSION >= 1025)
        if ((request->Type&MFX_MEMTYPE_VIDEO_MEMORY_ENCODER_TARGET) && (request->Type & MFX_MEMTYPE_INTERNAL_FRAME))
        {
            desc.BindFlags = D3D11_BIND_DECODER | D3D11_BIND_VIDEO_ENCODER;
        }
        else
#endif
            desc.BindFlags = D3D11_BIND_DECODER;
```

**Comment by jpark37** [comment]
Remaining issues:
- Burst of encoder overload at start of recording (known-shippable IMO)
- Add mfxExtMasteringDisplayColourVolume, mfxExtContentLightLevelInfo (if supported, bug Intel if not supported)
- Ask Intel why mfxExtChromaLocInfo is not supported, PQ/HLG should specify top-left
- Need to test config permutations succeed/fail as expected; encoder fallbacks

**Review by jpark37 — APPROVED** [review]

**Comment by ma3uk** [comment]
If the QSV encoder is being updated, is it possible to add additional encoding parameters such as -bref, -deblock_alpha, -deblock_beta, -x (number of reference frames), -ppyr:<on,off> for H264 and H265?

**Comment by ma3uk** [comment]
@yuriy-chumak  @jpark37  If the QSV encoder is being updated, is it possible to add additional encoding parameters such as -bref, -dblk_alpha, -dblk_beta, -x (number of reference frames), -ppyr:<on, off>, -AdaptiveI:<on, off>, -AdaptiveB:<on, off> for H264 and H265, so that they are available at least in the line of simple mode?

**Comment by yuriy-chumak** [comment]
@ma3uk,
> is it possible to add additional encoding parameters

It's possible, sure.
I'm not an obs-studio maintainer, just a thirdparty coder with some encoding knowledge. And my opinion is "you should create a [feature request](https://github.com/obsproject/obs-studio/issues) with a reasoned description of why you need this feature". This request will be visible to much more people and some of them probably implement it.

File: plugins/obs-qsv11/obs-qsv11-plugin-main.c
Comment (brittneysclark) [inline plugins/obs-qsv11/obs-qsv11-plugin-main.c:69]: We suggest only enabling HEVC for iGPUs from 10th Gen platforms (ICL+) and Arc graphics. A combination of CPU ID check for iGPU cases on 10th Gen+ and MFXVideoCORE_QueryPlatform to check mfxPlatform->MediaAdapterType can help detect Arc graphics.

Comment (jpark37) [reply plugins/obs-qsv11/obs-qsv11-plugin-main.c:69]: How about something like this?
```
bool allow_hevc =
    (media == MFX_MEDIA_DISCRETE) ||
    ((media == MFX_MEDIA_INTEGRATED) && (platform >= MFX_PLATFORM_ICELAKE);
```

Comment (jpark37) [reply plugins/obs-qsv11/obs-qsv11-plugin-main.c:69]: I updated the code to check inside qsv_encoder_open. Resolve the conversation if you agree.

Comment (brittneysclark) [reply plugins/obs-qsv11/obs-qsv11-plugin-main.c:69]: Did a few quick tests on my side - this looks good. There is no option for me to resolve the conversation though, only comment. Can you close it please?

**Comment by jpark37** [comment]
Setting back to draft until we know whether metadata is actually corrupt or not.

**Pull request merged**
